src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectWorkspaceStateGenerator.cs

```diff
@@ -4,133 +4,116 @@
 
-    // Used in unit tests to ensure we can know when background work finishes.
-    public ManualResetEventSlim? NotifyBackgroundWorkCompleted { get; set; }
+    private ManualResetEventSlim? _blockBackgroundWorkStart;
+    private ManualResetEventSlim? _notifyBackgroundWorkCompleted;
 
-    public Task UpdateAsync(Project? workspaceProject, IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    public void Dispose()
     {
-        if (projectSnapshot is null)
+        if (_disposeTokenSource.IsCancellationRequested)
         {
-            throw new ArgumentNullException(nameof(projectSnapshot));
+            return;
         }
 
-        return _dispatcher.RunAsync(
-            () =>
-            {
-                if (_disposed)
-                {
-                    return;
-                }
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
 
-                if (Updates.TryGetValue(projectSnapshot.Key, out var updateItem) &&
-                    !updateItem.Task.IsCompleted &&
-                    !updateItem.Cts.IsCancellationRequested)
-                {
-                    updateItem.Cts.Cancel();
-                }
+        foreach (var (_, updateItem) in _updates)
+        {
+            updateItem.Dispose();
+        }
 
-                if (updateItem?.Cts.IsCancellationRequested == false)
-                {
-                    updateItem?.Cts.Dispose();
-                }
+        // Release before dispose to ensure we don't throw exceptions from the background thread trying to release
+        // while we're disposing. Multiple releases are fine, and if we release and it lets something passed the lock
+        // our cancellation token check will mean its a no-op.
+        _semaphore.Release();
+        _semaphore.Dispose();
 
-                var lcts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-                var updateTask = Task.Factory.StartNew(
-                    () => UpdateWorkspaceStateAsync(workspaceProject, projectSnapshot, lcts.Token),
-                    lcts.Token,
-                    TaskCreationOptions.None,
-                    TaskScheduler.Default).Unwrap();
-                updateItem = new UpdateItem(updateTask, lcts);
-                Updates[projectSnapshot.Key] = updateItem;
-            },
-            cancellationToken);
+        _blockBackgroundWorkStart?.Set();
     }
 
-    public void Dispose()
+    public void EnqueueUpdate(Project? workspaceProject, IProjectSnapshot projectSnapshot)
     {
-        if (_disposed)
+        if (_disposeTokenSource.IsCancellationRequested)
         {
             return;
         }
 
-        _disposed = true;
-
-        foreach (var update in Updates)
+        lock (_updates)
         {
-            if (!update.Value.Task.IsCompleted &&
-                !update.Value.Cts.IsCancellationRequested)
+            if (_updates.TryGetValue(projectSnapshot.Key, out var updateItem))
```

Comment:
Are updates ever removed from the dictionary? It looks like only when `CancelUpdates` is called

Reply:
Correct, and I don't love that. However, I'm a little nervous about changing the behavior to remove items after an update occurs.

Reply:
That's a good idea, but I'm not seeing conditions where it'd cause false noise. Could you elaborate?

Reply:
AFAICT, it'll only log if there was a previous update that was already cancelled immediately before it replaces that update. And, because `_updates` is protected by a lock, we shouldn't ever be logging for the same update that was previously cancelled more than once.

Reply:
Is that because the `projectSnapshot.Key` is different across updates?

Reply:
`projectSnapshot.Key` isn't different across updates. It should be the same for each snapshot of the same project (so long as the intermediate output path doesn't change). However, I still don't see why there'd be false noise.

Reply:
Oh wait, I see. We'd get false noise if the update completed and wasn't cancelled and we're called to update the project again.

Reply:
`_logger.LogTrace($"Cancelling previously enqueued update for '{projectSnapshot.FilePath}'.");` 

That reports cancelling even if it's already completed the task right? That's what I'm saying, we'll always say we're cancelling after the first update, which seems to be virtually useless logging because it may have finished the update

Reply:
Yes - thank you! I just wasn't seeing it.